### PR TITLE
fix(qwik-city): support vite.base in dev mode

### LINKED ISSUE #5984: [🐞] qwik-city: vite base doesn't work in dev mode
### Which component is affected?

Qwik City (routing)

### Describe the bug

When you set `config.base` of vite, the build correctly uses it but dev mode breaks, getting module files without the base path.

### Reproduction

https://stackblitz.com/edit/qwik-starter-wvwual?file=vite.config.ts

### Steps to reproduce

try clicking Home

### System Info

```shell
stackblitz, 1.5.1
```

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by JerryWu1234**
Fixes #5984

### CODE REVIEW

File: packages/qwik/src/qwikloader.ts
Comment (wmertens) [inline packages/qwik/src/qwikloader.ts:77]: what is this for? Aren't all qrls hosted in the same place?

Comment (wmertens) [inline packages/qwik/src/qwikloader.ts:77]: what is this for? Aren't all qrls hosted in the same place?

**Comment by JerryWu1234** [comment]
@wmertens didn't finish yet.
I create PR just for convenient sync code. lol

**Comment by JerryWu1234** [comment]
@wmertens I'm sorry to bother you, I know where problem is, but I was stuck on a crucial problem.

please take a look at [this line ](https://github.com/BuilderIO/qwik/blob/9959b1875ac2b5f70de28e3f97baf86bda0a56c6/packages/qwik/src/server/utils.ts#L30)

Why does it set a default `'/build/'`,
and what's between `q:route` and `q:base`.

**Comment by wmertens** [comment]
/build is where the scripts are stored by default. The result of this function should then get the vite base added.

**Comment by JerryWu1234** [comment]
image
image

@wmertens 
I have almost done.
because it's different between pro and dev.
pro `q:base="/newbase/build/"`
dev `q:base="build/"`

I want to check the pro env, but code is compressed by the tool. when I run `pnpm preview`
Could you give me some tips? 
thanks

**Comment by wmertens** [comment]
Btw, you can set minify:false in the build to see prod as-is

**Comment by JerryWu1234** [comment]
yR6clyXv70
pro

mLx3DdTwLv
dev

everything is right

**Comment by JerryWu1234** [comment]
I also tested env and pro without base added

File: packages/qwik/src/optimizer/src/plugins/plugin.ts
Comment (wmertens) [inline packages/qwik/src/optimizer/src/plugins/plugin.ts:139]: What if base is 'a'? Also, you default it to '/' and then replace it with ''

File: packages/qwik/src/optimizer/src/plugins/plugin.ts
Comment (wmertens) [inline packages/qwik/src/optimizer/src/plugins/plugin.ts:968]: Perhaps you should rename this to baseNoSlash because the vite base requires a slash.

Or else make it include the slash so it's the same and then handle the slash when using it

File: packages/qwik/src/optimizer/src/plugins/vite-server.ts
Comment (wmertens) [inline packages/qwik/src/optimizer/src/plugins/vite-server.ts:121]: Isn't it possible that done of these urls already include the base?

Better to change them at the source, no?

File: packages/qwik/src/optimizer/src/plugins/vite.ts
Comment (wmertens) [inline packages/qwik/src/optimizer/src/plugins/vite.ts:158]: This one includes a slash

File: packages/qwik/src/qwikloader.ts
Comment (wmertens) [inline packages/qwik/src/qwikloader.ts:78]: Same comments as above

**Review by wmertens — COMMENTED** [review]
Great, but we need some unit tests and more consistent `/` handling

**Comment by JerryWu1234** [comment]
> Great, but we need some unit tests and more consistent `/` handling

I'll add test after you reviewed.

**Comment by wmertens** [comment]
@JerryWu1234 how's it going with the tests? Need help?

**Comment by JerryWu1234** [comment]
> @JerryWu1234 how's it going with the tests? Need help?



> @JerryWu1234 how's it going with the tests? Need help?

I haven't tried to add tests yet because I’m waiting for your reply on the problem you mentioned

**Comment by wmertens** [comment]
@JerryWu1234 which one specifically? I don't have any more information to add right now?

**Comment by JerryWu1234** [comment]
> @JerryWu1234 which one specifically? I don't have any more information to add right now?

https://github.com/BuilderIO/qwik/pull/6017#discussion_r1533292949

https://github.com/BuilderIO/qwik/pull/6017#discussion_r1533286154

Here

**Comment by wmertens** [comment]
So I meant that we should have unit tests, and that you shouldn't have to replace the double slashes if you don't put them in.

So you need to be careful about base, don't call replace, and have unit tests that prove it doesn't add double slashes for various test cases

**Comment by JerryWu1234** [comment]
> So I meant that we should have unit tests, and that you shouldn't have to replace the double slashes if you don't put them in.
> 
> So you need to be careful about base, don't call replace, and have unit tests that prove it doesn't add double slashes for various test cases



> So I meant that we should have unit tests, and that you shouldn't have to replace the double slashes if you don't put them in.
> 
> So you need to be careful about base, don't call replace, and have unit tests that prove it doesn't add double slashes for various test cases

Let me try to add some tests

**Comment by JerryWu1234** [comment]
> @JerryWu1234 how's it going with the tests? Need help?

@wmertens 

I really need your help.

To test the base attribute, we need to rely on both **build** and **start**. However, the current e2e only has build.

If I modify it, I'm afraid it will break the project's testing structure.

**Comment by wmertens** [comment]
@JerryWu1234 🤔 actually the tests in v2 are a lot better. Take a look at the build/V2 branch, do you think you can incorporate your test there?

And actually, maybe that's not needed. It could be enough to make unit tests, not e2e

**Comment by JerryWu1234** [comment]
> @JerryWu1234 🤔 actually the tests in v2 are a lot better. Take a look at the build/V2 branch, do you think you can incorporate your test there?
> 
> And actually, maybe that's not needed. It could be enough to make unit tests, not e2e

https://github.com/BuilderIO/qwik/blob/87f0b9e5acf89f03204cc8e1cddaca8f54ad8ee9/starters/playwright.config.ts#L28
https://github.com/BuilderIO/qwik/blob/87f0b9e5acf89f03204cc8e1cddaca8f54ad8ee9/starters/dev-server.ts#L63
The build/v2 branch does not seem to have many changes compared to the master branch. It also seems that it is not possible to dynamically pass in the base parameter.

Here is my idea:

I referenced the [Astro](https://github.com/withastro/astro/blob/57959e607e81ec3b13bc55db6daa253a63485a97/packages/astro/test/test-utils.js#L97) test code, and I think I can dynamically pass in the **base** parameter in both **build** and **start** modes. **Vitest** can cover most scenarios, but I don't know how to test the **qwikLoader** function in Vitest, , so I want to use **e2e** tests.

**Comment by wmertens** [comment]
> I think I can dynamically pass in the base parameter in both build and start modes

sounds good, give it a shot!

**Comment by wmertens** [comment]
@JerryWu1234 there's a TS error https://github.com/BuilderIO/qwik/actions/runs/8508365123/job/23414373840?pr=6017#step:7:14

**Comment by JerryWu1234** [comment]
@wmertens 

1. Could you please evaluate the feasibility of my test plan? 
2. I can't write a complete test yet because there is an error. The error is a bit complicated. I hope you can help me take a look.
image

**Review by wmertens — CHANGES_REQUESTED** [review]
I propose you leave the tests out, on the v2 branch it will be clearer how to test e2e

**Comment by wmertens** [comment]
@JerryWu1234 actually the e2e tests don't test qwik-city yet in v2, so we'll have to wait for that to add a test

**Comment by JerryWu1234** [comment]
> actually the e2e tests don't test qwik-city yet in v2, so we'll have to wait for that to add a test

ok . add tests late

**Comment by JerryWu1234** [comment]
last check before merge 

dev
has /newbase/
image


no /newbase/
image

pro
has /newbase/
image

no /newbase/
image

**Comment by PatrickJS** [comment]
I do need this but I do set `base` to a cloudfront url so the regexp may not work

**Comment by JerryWu1234** [comment]
> I do need this but I do set `base` to a cloudfront url so the regexp may not work

good idea
I will cover this function

**Comment by JerryWu1234** [comment]
image
image
@PatrickJS 
it has already covered it

**Comment by JerryWu1234** [comment]
@wmertens When will **build/v2** branch finish the test enhancement ?

**Comment by JerryWu1234** [comment]
@wmertens is there anything else I need to do in this PR? 
It will encourage me to keep going if this PR is merged.
thanks

File: packages/qwik/src/core/platform/platform.ts
Comment (wmertens) [inline packages/qwik/src/core/platform/platform.ts:70]: `url` can be anything, including query string and hash, so you shouldn't run replace on it.
For example `toUrl(doc, el, "/hello/foo?stuff=///meep#123")` should work.
Just check if url isn't empty and starts with /

File: packages/qwik/src/core/platform/platform.ts
Comment (wmertens) [inline packages/qwik/src/core/platform/platform.ts:68]: not sure why this is needed?

File: packages/qwik/src/core/platform/platform.ts
Comment (wmertens) [inline packages/qwik/src/core/platform/platform.ts:71]: .origin can be `null`, for example when using Capacitor which uses a base of `capacitor://`

**Review by wmertens — CHANGES_REQUESTED** [review]
I propose that you make sure that everywhere, `base` ends with a `/`. If you need to have it without `/` somewhere, slice it.

Also keep in mind that urls can be any url with any schema.

**Comment by JerryWu1234** [comment]
I done

**Comment by JerryWu1234** [comment]
@wmertens

**Comment by PatrickJS** [comment]
@JerryWu1234 can you fix the conflicts

**Comment by JerryWu1234** [comment]
> @JerryWu1234 can you fix the conflicts

Will fix it today

**Comment by wmertens** [comment]
wait, actually, it's only pnpm-lock.yaml, so you can `git checkout main -- pnpm-lock.yaml && pnpm i`

**Comment by JerryWu1234** [comment]
image

@wmertens I done this

**Comment by JerryWu1234** [comment]
but still changed yaml file

**Comment by JerryWu1234** [comment]
I copied code directly from the main branch

File: starters/apps/basic/src/root.tsx
Comment (PatrickJS) [inline starters/apps/basic/src/root.tsx:24]: does BASE_URL end with a /?

**Review by PatrickJS — CHANGES_REQUESTED** [review]

File: starters/apps/basic/src/root.tsx
Comment (PatrickJS) [reply starters/apps/basic/src/root.tsx:24]: @JerryWu1234 one question then I can merge

Comment (PatrickJS) [reply starters/apps/basic/src/root.tsx:24]: I'm going to merge and we can figure out later (at least we have time before we cut a release)

**Comment by PatrickJS** [comment]
e2e tests are failing 🤔

**Comment by JerryWu1234** [comment]
image
image

it's weird, I didn't modify entry.ssr.ts. however, I have to dive into

**Comment by PatrickJS** [comment]
yeah very weird

**Comment by PatrickJS** [comment]
oh maybe its erroring since the e2e tests are running the app in a browser so if the files can't load. thats the only thing i can think of

**Comment by PatrickJS** [comment]
idk about this I think we need to make sure /build/ is added for scripts even if base was set

**Comment by wmertens** [comment]
I think we need to take a step back here. `q:base` is the absolute path or full URL of where Qwik is supposed to get the dist files, and vite `base` is the prefix path to the url of the application.

- We need to make sure we use the right thing at the right time.
- q:base needs to be absolute or a full URL and end with `/`. It needs to include the vite base if it's not explicitly set.
- vite base needs to be absolute and end with `/`

I see a bunch of duplication around / handling, we can't have that. It needs to be simple throughout, with a helper function if needed. We know that the bases will always end with `/`.

Furthermore, it's up to the dev to use the vite base in their URLs, we can't add it for them, because there might be links between microfrontends.

**Comment by JerryWu1234** [comment]
> I think we need to take a step back here. `q:base` is the absolute path or full URL of where Qwik is supposed to get the dist files, and vite `base` is the prefix path to the url of the application.
> 
> * We need to make sure we use the right thing at the right time.
> * q:base needs to be absolute or a full URL and end with `/`. It needs to include the vite base if it's not explicitly set.
> * vite base needs to be absolute and end with `/`
> 
> I see a bunch of duplication around / handling, we can't have that. It needs to be simple throughout, with a helper function if needed. We know that the bases will always end with `/`.
> 
> Furthermore, it's up to the dev to use the vite base in their URLs, we can't add it for them, because there might be links between microfrontends.

just want to confirm I understand or not.
such as if user set a Vite base: '/newbase/'
so that the `q:base` is `http://localhost:3000/newbase/` in dev 
and the `q:base` is `http://localhost:3000/newbase/build/` in pro 

@PatrickJS @wmertens 

I think we can try to easy way to fix this bug and waiting for built/v2.
because it's very difficult to guarantee whether it's  right or not without tests

**Comment by PatrickJS** [comment]
q:base is for assets so it's saying this is where the /build/ files are. I think it makes sense right now to always make sure /build/ is in the url

File: starters/apps/basic/src/root.tsx
Comment (sreeisalso) [reply starters/apps/basic/src/root.tsx:24]: It will be better to add / here instead of assuming BASE_URL ends with /, because even if it has a // will be treated as /

**Comment by wmertens** [comment]
Actually I wonder if https://github.com/QwikDev/qwik/commit/b381b228531badce0253a6341c27d65f1a8023c0 fixed this - can you check if 1.5.5 still needs this PR @JerryWu1234 ?

**Comment by JerryWu1234** [comment]
image
image
still has a problem

**Comment by wmertens** [comment]
Thank you for checking @JerryWu1234, can you figure out exactly what is wrong? Is the q:base wrong, or is it requesting specific files wrongly etc

**Comment by JerryWu1234** [comment]
> Thank you for checking @JerryWu1234, can you figure out exactly what is wrong? Is the q:base wrong, or is it requesting specific files wrongly etc

Just check where wrong is? or fix it in this PR?

**Comment by wmertens** [comment]
@JerryWu1234 yes just where. Is it by any chance when it's trying to load `/src/something..._hashhashhash.js`? Then I know where to fix it (and I'm replacing that anyway)

**Comment by PatrickJS** [comment]
I'll fix the pr

**Comment by JuanmaVargas** [comment]
In production, is there any workaround to fix this? I'm trying to create a multi-zone in Next.js for a Qwik application, and for that, I need the vite.base set in the configuration to work for all the assets

**Comment by PatrickJS** [comment]
I'll revert my changes except for the important ones and update the pr

**Comment by PatrickJS** [comment]
@JuanmaVargas can you use [this](https://github.com/QwikDev/qwik/pull/6017#issuecomment-2155490844) to test out the changes

**Comment by wmertens** [comment]
@JerryWu1234 thank you for your efforts, but I think it's fixed now in 1.7.1 (in a different way) so I'll close this.

**Pull request closed without merging**
